algo/src/test/java/org/neo4j/graphalgo/impl/UnionFindSafetyTest.java

```diff
@@ -0,0 +1,230 @@
+import org.neo4j.graphalgo.api.RelationshipIntersect;
+import org.neo4j.graphalgo.api.WeightedRelationshipConsumer;
+import org.neo4j.graphalgo.core.utils.Pools;
+import org.neo4j.graphalgo.core.utils.paged.AllocationTracker;
+import org.neo4j.graphdb.Direction;
+import org.neo4j.helpers.Exceptions;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Random;
+import java.util.Set;
+import java.util.function.LongPredicate;
+import java.util.stream.IntStream;
+
+import static org.junit.Assert.assertSame;
+
+@RunWith(Parameterized.class)
+public final class UnionFindSafetyTest {
+
+    @Parameterized.Parameters(name = "{0}")
+    public static Collection<Object[]> data() {
+        return Arrays.asList(
+                new Object[]{UnionFindAlgo.QUEUE},
+                new Object[]{UnionFindAlgo.FORK_JOIN},
+                new Object[]{UnionFindAlgo.FJ_MERGE}
+        );
+    }
+
+    @Parameterized.Parameter
+    public UnionFindAlgo unionFindAlgo;
+
+    @Test(timeout = 2000L)
+    public void testUnionFindSafetyUnderFailure() {
+        IllegalStateException error = new IllegalStateException("some error");
+        Graph graph = new FlakyGraph(100, 10, new Random(42L), error);
+        try {
+            unionFindAlgo.run(
+                    graph,
+                    Pools.DEFAULT,
+                    10,
+                    10
+            );
+        } catch (Throwable e) {
+            assertSame(error, Exceptions.rootCause(e));
+        }
+    }
+
+    @Test(timeout = 2000L)
+    public void testHugeUnionFindSafetyUnderFailure() {
+        IllegalStateException error = new IllegalStateException("some error");
+        HugeGraph graph = new FlakyGraph(100, 10, new Random(42L), error);
+        try {
+            unionFindAlgo.run(
+                    graph,
+                    Pools.DEFAULT,
+                    AllocationTracker.EMPTY,
+                    10,
+                    10
+            );
+        } catch (Throwable e) {
+            assertSame(error, Exceptions.rootCause(e));
+        }
+    }
+
+    private static final class FlakyGraph implements HugeGraph {
+        private final int nodes;
+        private final int maxDegree;
+        private final Random random;
+        private final RuntimeException error;
+
+        private FlakyGraph(int nodes, int maxDegree, Random random, RuntimeException error) {
+            this.nodes = nodes;
+            this.maxDegree = maxDegree;
+            this.random = random;
+            this.error = error;
+        }
+
+        @Override
+        public void canRelease(final boolean canRelease) {
+        }
+
+        @Override
+        public long nodeCount() {
+            return (long) nodes;
+        }
+
+        @Override
+        public void forEachRelationship(
+                final long nodeId,
+                final Direction direction,
+                final HugeRelationshipConsumer consumer) {
+            if (nodeId == 0L) {
```

Comment:
why exactly at node 0 ?
i.e. at the beginning ...

Reply:
It's an arbitrary decision where I know that this will happen at least once. We could also always fail or randomly.